src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.GroupedAnalyzerActionsForAnalyzer.cs

```diff
@@ -70,14 +81,17 @@ public ImmutableSegmentedDictionary<TLanguageKindEnum, ImmutableArray<SyntaxNode
                 {
                     if (_lazyNodeActionsByKind == null)
                     {
-                        var nodeActions = _analyzerActionsNeedFiltering ?
-                            AnalyzerActions.GetSyntaxNodeActions<TLanguageKindEnum>(_analyzer) :
-                            AnalyzerActions.GetSyntaxNodeActions<TLanguageKindEnum>();
+                        var nodeActions = ArrayBuilder<SyntaxNodeAnalyzerAction<TLanguageKindEnum>>.GetInstance();
+                        if (_analyzerActionsNeedFiltering)
+                            AnalyzerActions.AddSyntaxNodeActions(_analyzer, nodeActions);
+                        else
+                            AnalyzerActions.AddSyntaxNodeActions(nodeActions);
+
                         VerifyActions(nodeActions, _analyzer);
-                        var analyzerActionsByKind = !nodeActions.IsEmpty ?
-                            AnalyzerExecutor.GetNodeActionsByKind(nodeActions) :
-                            ImmutableSegmentedDictionary<TLanguageKindEnum, ImmutableArray<SyntaxNodeAnalyzerAction<TLanguageKindEnum>>>.Empty;
+                        var analyzerActionsByKind = AnalyzerExecutor.GetNodeActionsByKind(nodeActions);
```

Comment:
swithcing to a temp array builder allows us to only accept ArrayBuilders in GetNodeActionsByKind instead of needing to take an IEnumerable ot handle ArrayBuilders or ImmutableArrays.  This also avoids intermediary copies of data.